cmd/object-utils.go

```diff
@@ -163,28 +163,27 @@ func (d byBucketName) Len() int           { return len(d) }
 func (d byBucketName) Swap(i, j int)      { d[i], d[j] = d[j], d[i] }
 func (d byBucketName) Less(i, j int) bool { return d[i].Name < d[j].Name }
 
-// Copied from io.LimitReader()
-// limitReader returns a Reader that reads from r
-// but returns error after n bytes.
-// The underlying implementation is a *LimitedReader.
-type limitedReader struct {
-	R io.Reader // underlying reader
-	M int64     // min bytes remaining
-	N int64     // max bytes remaining
+// rangeReader returns a Reader that reads from r
+// but returns error after Max bytes read as errDataTooLarge.
+// but returns error if reader exits before reading Min bytes
+// errDataTooSmall.
+type rangeReader struct {
+	bucket, object string
+	Reader         io.Reader // underlying reader
+	Min            int64     // min bytes remaining
+	Max            int64     // max bytes remaining
 }
 
-func limitReader(r io.Reader, m, n int64) io.Reader { return &limitedReader{r, m, n} }
-
-func (l *limitedReader) Read(p []byte) (n int, err error) {
-	n, err = l.R.Read(p)
-	l.N -= int64(n)
-	l.M -= int64(n)
-	if l.N < 0 {
+func (l *rangeReader) Read(p []byte) (n int, err error) {
+	n, err = l.Reader.Read(p)
+	l.Max -= int64(n)
+	l.Min -= int64(n)
+	if l.Max < 0 {
 		// If more data is available than what is expected we return error.
-		return 0, errDataTooLarge
+		return 0, toObjectErr(errDataTooLarge, l.bucket, l.object)
```

Comment:
just errDataTooLarge should be returned here (similar to how streaming v4 returns just errSignatureMismatch without bucket/object info) the caller will take care of bucket/object related info